v3/src/liftoff/ref.ts

```diff
@@ -0,0 +1,127 @@
+import { setLocation, getLocation } from './loc'
+import { keyed } from './key'
+import { def, Bond, trace } from './pattern'
+
+export interface ScalarType<T> {
+  <X extends T=T>(tag: TemplateStringsArray, ...deps: any[]): (defaultValue?: X | Ref<X>) => Scalar<X>
+}
+
+export const DEFAULT_VALUE = Symbol('Default value, if any // Ref<T>::[DEFAULT_VALUE]: T | Ref<T>')
+export const DEFAULT_BOND = Symbol('Bond for default value // Bond')
+export interface Ref<T> {
+  // [DEFAULT_VALUE] value holds the default value for this ref
+  [DEFAULT_VALUE]: T | Ref<T> | void
+
+  // [DEFAULT_BOND] holds a bond which applies to this ref and binds
+  // the [DEFAULT_VALUE]. Creating this once is useful because it lets us
+  // set location data for the definition to the source location where the
+  // ref was defined.
+  [DEFAULT_BOND]: Bond
+}
+
+export interface Scalar<T> extends Ref<T> {
+  /**
+   * Provide a definition for this ref.
+   *
+   * @param value {T | Ref<T>} the definition
+   */
+  (value: T | Ref<T>): void
+  (tag: TemplateStringsArray, ...deps: any[]): (value: T | Ref<T>) => void
+
+  /**
+   * Provide a definition for this ref.
+   *
+   * @param value {T | Ref<T>} the definition
+   */
+  def(value: T | Ref<T>): void
+  def(tag: TemplateStringsArray, ...deps: any[]): (value: T | Ref<T>) => void
+}
+
+
+/**
+ * Create a scalar type, which can be called to create scalar refs.
+ */
+function createScalarType<T>
+  (tag: TemplateStringsArray, ..._deps: any[]): ScalarType<T> {
+    setLocation(tag, 2)
+    setLocation(createRef, tag)
+    const typeLabel = tag.join('___')
```

Comment:
I think it might be worth putting a limit on the length of the label here in order to keep it tamed for potential (future) serialization, metric reporting, visualization.  e.g. 255 max, but maybe even something much lower like 64.  Thoughts?

Reply:
Hmm. Maybe we could shorten it later, as needed for those uses?